cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTableAdmin.java

```diff
@@ -371,13 +342,19 @@ private StructuredTableSchema loadSchema(StructuredTableId tableId)
 
         fields.add(new FieldType(columnName, fromSQLType(resultSet.getString("data_type"))));
 
-        // "t" as true indicates primary key index, "f" as false indicates normal index, null means it's just a column
-        if ("t".equalsIgnoreCase(indexType)) {
-          primaryKeys.add(columnName);
-        } else if ("f".equalsIgnoreCase(indexType)) {
+        // "t" as true indicates primary key index
+        // "f" as false indicates normal index, but it can still be a primary key field that is part of other index
+        // null means it's just a column
+        if ("f".equalsIgnoreCase(indexType)) {
           indexes.add(columnName);
         }
       }
+
+      // Calling MetaData.getPrimaryKeys() to construct table primary keys and maintain the order
+      ResultSet rs = connection.getMetaData().getPrimaryKeys(null, null, tableId.getName());
```

Comment:
I believe the whole idea of the schemaStatement above was to get all data in one query. That query can be enhanced to add PK field order. No need to do another SQL

Reply:
That is true. However the `ordinal_position` in information_schema is by columns at DB creation time. One query is possible but need more complexed query statement. More investigation is needed on it

And it's not too bad to have one more query while loading schema from DB: 1. It's PK, it's currently immutable in CDAP system; 2. It's one time thing when the instance launch

Reply:
There is no "instance launch" thing, I believe it's something that runs on every pod start, so it's actually pretty important

Reply:
I believe the query change would be relatively simple. You would need to add something like `array_position(i.indkey, a.attnum)` and return it as a separate column (say `pk_order`). Then you use it so resort on java site

Reply:
brilliant! It works. I've pushed the chaneg